firmware/src/core/configmanager/ConfigManager.h

```diff
@@ -0,0 +1,88 @@
+#ifndef CONFIG_MANAGER_H
+#define CONFIG_MANAGER_H
+
+#include "Button.h"
+#include "Utils.h"
+#include "WifiManagerCustomParameters.h"
+#include "config_helper.h"
+#include "nvs_flash.h"
+#include <Preferences.h>
+#include <WiFiManager.h>
+#include <functional>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#define CM_PARAM_TYPE_STRING 0
+#define CM_PARAM_TYPE_INT 1
+#define CM_PARAM_TYPE_BOOL 2
+#define CM_PARAM_TYPE_FLOAT 3
+#define CM_PARAM_TYPE_COLOR 4
+#define CM_PARAM_TYPE_COMBOBOX 5
+
+// Uncomment to show debug output
+// #define CM_DEBUG
+
+class ConfigManager {
+public:
+    ConfigManager(WiFiManager &wm);
+    ~ConfigManager();
+
+    static ConfigManager *getInstance();
+
+    void setupWebPortal();
+    void saveAllConfigs();
+
+    // Register different types of variables with descriptions
+    void addConfigString(const std::string &section, const std::string &varName, std::string *var, size_t length, const std::string &description);
+    void addConfigInt(const std::string &section, const std::string &varName, int *var, const std::string &description);
+    void addConfigFloat(const std::string &section, const std::string &varName, float *var, const std::string &description);
+    void addConfigIP(const std::string &section, const std::string &varName, IPAddress *var, const std::string &description);
+    void addConfigBool(const std::string &section, const std::string &varName, bool *var, const std::string &description);
+    void addConfigColor(const std::string &section, const std::string &varName, int *var, const std::string &description);
+    void addConfigComboBox(const std::string &section, const std::string &varName, int *var, String options[], int numOptions, const std::string &description);
+
+    // Get stored values
+    String getConfigString(const std::string &varName, String defaultValue);
+    bool getConfigBool(const std::string &varName, bool defaultValue);
+    int getConfigInt(const std::string &varName, int defaultValue);
+    float getConfigFloat(const std::string &varName, float defaultValue);
+
+    // Register callbacks for changes
+    void addOnChangeCallback(
+        const std::string &section,
+        const std::string &varName,
+        const std::function<void(const std::string &section, const std::string &varName)> &callback);
+    void addOnChangeCallback(
+        const std::string &section,
+        const std::function<void(const std::string &section, const std::string &varName)> &callback);
+
+    // Check if a restart is required
+    bool isRequiresRestart() { return requiresRestart; }
+
+private:
+    struct Parameter {
```

Comment:
This thing really needs a destructor, because you allocate your `section` and `variableName`,
and those will need to be freed. (`free[] section; free[] variableName`)

Reply:
Actually, I changed the ConfigManager class to use const char* directly for section/varName/description and not create buffers anymore. This also helps to reduce the memory footprint a bit. And we do not need to free or delete anything here, I think.

The only downside is that the parameters now really need to be const, but that should be the case for these three.